src/probnum/filtsmooth/particlefiltsmooth/_particle_filter.py

```diff
@@ -0,0 +1,219 @@
+        )
+        rvs = [curr_rv]
+
+        for idx in range(1, len(times)):
+            curr_rv, _ = self.filter_step(
+                start=times[idx - 1],
+                stop=times[idx],
+                randvar=curr_rv,
+                data=dataset[idx],
+            )
+            rvs.append(curr_rv)
+        return ParticleFilterPosterior(rvs, times)
+
+    def filter_step(self, start, stop, randvar, data):
+        """Perform a particle filter step.
+
+        It consists of the following steps:
+        1. Propagating the "past" particles through the dynamics model.
+        2. Computing a "proposal" random variable.
+        This is either the prior dynamics model or the output of a filter step
+        of an (approximate) Gaussian filter.
+        3. Sample from the proposal random variable. This is the "new" particle.
+        4. Propagate the particle through the measurement model.
+        This is required in order to require the PDF of the resulting RV at
+        the data. If this is small, the weight of the particle will be small.
+        5. Compute weights ("event probabilities") of the new particle.
+        This requires evaluating the PDFs of all three RVs (dynamics, proposal, measurement).
+
+        After this is done for all particles, the weights are normalized in order to sum to 1.
+        """
+        new_weights = randvar.probabilities.copy()
+        new_support = randvar.support.copy()
+
+        for idx, (particle, weight) in enumerate(zip(new_support, new_weights)):
+
+            dynamics_rv, _ = self.dynamics_model.forward_realization(
+                particle, t=start, dt=(stop - start)
+            )
+            proposal_state, proposal_weight = self.compute_new_particle(
+                data, stop, dynamics_rv
+            )
+            new_support[idx] = proposal_state
+            new_weights[idx] = proposal_weight
+
+        new_weights = new_weights / np.sum(new_weights)
+        new_rv = random_variables.Categorical(
+            support=new_support,
+            probabilities=new_weights,
+            random_state=self.random_state,
+        )
+
+        if self.with_resampling:
+            if effective_number_of_events(new_rv) < self.min_effective_num_of_particles:
+                new_rv = new_rv.resample()
+
+        return new_rv, {}
+
+    def compute_new_particle(self, data, stop, dynamics_rv):
+        """Compute a new particle.
+
+        Turn the dynamics RV into a proposal RV, apply the measurement
+        model and compute new weights via the respective PDFs.
+        """
+        proposal_rv = self.dynamics_to_proposal_rv(dynamics_rv, data=data, t=stop)
+        proposal_state = proposal_rv.sample()
+        meas_rv, _ = self.measurement_model.forward_realization(proposal_state, t=stop)
+
+        log_proposal_weight = (
+            meas_rv.logpdf(data)
+            + dynamics_rv.logpdf(proposal_state)
+            - proposal_rv.logpdf(proposal_state)
```

Comment:
Am I understanding correctly, that for the bootstrap PF, this just cancels?

Reply:
yes. I add an if-else here, to make this point even more obvious.